### DISCUSSION TITLE
How to install fortls in offline windows

### DISCUSSION THREAD

**Discussion by lvkas521424** [fortran-lang/fortls#434] [General]
Hello ,everyone,
Excuse me, may I ask which packages my Python should have to install fortls on the Windows system on the intranet? It cannot be installed directly using `pip install fortls`.

**Comment by gnikit** [comment]
There's a few ways that this can be done, but essentially, on a system with internet you'll have to download the dependencies of fortls

- https://github.com/fortran-lang/fortls/blob/3cf8f2978b470bbffb4bbd87d242c59d8b9f5709/pyproject.toml#L41-L42
- and then the Python [wheels]( https://files.pythonhosted.org/packages/84/dd/45b6372aba0217cba3168de33ac562352e1d3f7a3007369a3dbaa5aa07e2/fortls-3.1.2-py3-none-any.whl)

Send them to the offline machine and then run pip install. You might find [this](https://stackoverflow.com/questions/11091623/how-to-install-packages-offline) SO post useful

**Reply by lvkas521424** [reply]
Thanks, I having install it in this way!  ^_^

**Reply by TLCFEM** [reply]
@gnikit 
Do you consider releasing binary files of `fortls`?
It can be easily packed using for example `pyinstaller` so that there is no further dependency on a valid python env.

**Reply by gnikit** [reply]
This is something I started seriously looking into the other day, because I am tired of all the complications and the support queries that arise from the Modern Fortran from VS Code, but I haven't finished looking into all the alternatives (PyInstaller, cx_Freeze` Nuitka, or PyOxidizer).

The problem is this would only solve issues with fortls. I would have to repeat the same work for findent and fprettify (and pretty much any other program used by Modern Fortran).
Currently, I am rewriting the Python manager in the Modern Fortran extension so that I make things a bit easier to maintain on that front.

Also, in terms of standalone binaries, I am not entirely sure how I would distribute them. GitHub is a good place but not everyone has a GitHub account. 

So in short, yes, I am looking into it.

**Reply by TLCFEM** [reply]
My two cents:

The VS Code extension can bundle binaries like `fortls`, it can be done via CI automation, using `pyinstaller`, `nuitka`, or other tools of your liking.
By such, the extension does not execute `pip install fortls`, instead, it finds the binary and executes it, or falls back to use the bundled binaries, see for examle the `ruff` extension for VS Code.
Users, especially novices, do not need to install python and create a valid venv first.

> Also, in terms of standalone binaries, I am not entirely sure how I would distribute them. GitHub is a good place but not everyone has a GitHub account.

Not sure why a GitHub account is necessary here, but I presume if the binaries are publicly released, users can download the binaries, the extension/tool can `fetch` them, `pip install fortls` requires the same Internet access anyway.

**Reply by gnikit** [reply]
Yes, VS Code could bundle binaries, although there are serious security concerns in doing that. The problem is that you are swapping one type of problem with another. You want people to be using the latest stable version of `fortls` regardless of the version of the Modern Fortran extension. If you bundle binaries in the extension that's no longer possible. If you download binaries upon installation, that is possible but someone (probably me) will have to write the code for this and maintain the CD infrastructure.
One would also have to modify `fortls` in order to disable the autoupdate feature. There are quite a few more issues with using binaries of `fortls` by default in vscode, etc.

**Reply by TLCFEM** [reply]
Understood.
Some comments regarding your previous reply.

> You want people to be using the latest stable version of fortls regardless of the version of the Modern Fortran extension. If you bundle binaries in the extension that's no longer possible.

I do not see why it is not possible, it is feasible to check the bundled version and fetch new versions and prompt whether to update, if the bundled binary is in use, simply replace the existing one with the newer version.
For example, the `go` extension does not ship with its language server [`gopls`](https://github.com/golang/tools/tree/master/gopls), but installs it on activation.

It is also feasible to have an option to assign `fortls` path such that users can manually manage `fortls` binary location and update if needed.
The extension could fall back to use a bundled binary, or the installed python package in the current environment, etc.

> If you download binaries upon installation, that is possible but someone (probably me) will have to write the code for this and maintain the CD infrastructure.

I think I have the capacity doing so, if you want to offload to someone else, please let me know.

Publishing `fortls` binary in the `fortls` repository does not change the vs code logic ATM, it is a prerequisite though.
Maybe it is the very first thing to start with.
I can create PR for this if you like.

**Reply by gnikit** [reply]
I think I'll have to drive this ATM due to the permissions required/compliance concerns, but I'll tag you when I get around to it. 

If you still want to post some snippets of code/prs I'll gladly have a look in the next few months, but please don't spend too much time on this.
